In [1]:
import torch
from torch.autograd import Variable

Variable is a sort of wrapper around tensor that has several important functions built in.

In [2]:
x = Variable(torch.ones(2, 2), requires_grad=True)
print(x)

Variable containing:
 1  1
 1  1
[torch.FloatTensor of size 2x2]



It's behavior is similar to a tensor.

In [3]:
y = x + 2
print(y)

Variable containing:
 3  3
 3  3
[torch.FloatTensor of size 2x2]



Graphs are defined dynamically  in Torch whenever a variable builts a dependency on another variable/function of that variable. Since $y$ is created as a result of an operation on a variable it has a grad_fn that links back to the variable $x$. 


In [4]:
print(y.grad_fn)

Suppose we do a few operations on the variable $z$ and store it's mean in a new variable $out$.


In [5]:
z = y * y * 3
out = z.mean()
print(z, out)

Variable containing:
 27  27
 27  27
[torch.FloatTensor of size 2x2]
 Variable containing:
 27
[torch.FloatTensor of size 1]



We can now backpropagate the variable $out$ to get the gradient of $out$ with respect to the graph leaves i.e. with respect to it's parent root node (here $x$). The values of the gradient for $x$ are updated accorrdingly.

In [6]:
out.backward()
print(x.grad)

Variable containing:
 4.5000  4.5000
 4.5000  4.5000
[torch.FloatTensor of size 2x2]



Why do we get a $2\times2$ matrix of $4.5$s?

We get $4.5$ since $4.5$ is the value of $\frac{d}{dx}out$

Simple. each $z_i=3(x_i+2)^2$. Since all $x$ is same (and all $z$ is same) and $z = 3(x+2)^2$

Therefore $\frac{d}{dx}z_i = \frac{3}{2(x_i+2)}$ and Since $x_i = 1$, it is $\frac{3}{2\times3}$ = 4.5

<img src="http://pytorch.org/tutorials/_images/Variable.png">

Thus we see the <b>autograd</b> package is very useful and by defining Variables, we can access the automatic backpropagation function while also being able to use numpy-like syntax for processing our variables/matrices.